### PR TITLE
ASoC: SOF: Add DMIC token for unmute gain ramp time

### PR DESCRIPTION
**Pull request by singalsu**
The settling time of DMIC DC level is both platform and used
microphone model specific. The unmute gain ramp is used to conceal
most of the large DC level seen in beginning of capture. This patch
adds into the DMIC DAI IPC struct a new field called unmute_ramp_time
and a new token SOF_TKN_INTEL_DMIC_UNMUTE_RAMP_TIME. The value is the
ramp length in milliseconds (ms).

The ABI minor version is incremented for this backwards compatible
change.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

### CODE REVIEW

**Comment by singalsu** [comment]
ETA for corresponding PR for FW is next Friday (tomorrow is bank holiday). The change will be backwards compatible both ways. With new kernel and new FW the DMIC capture unmute ramp time will be topology defined.

**Comment by plbossart** [comment]
@lgirdwood need to agree on which ABI version we use for this change.

File: include/uapi/sound/sof/tokens.h [hunk 1]
```diff
@@ -85,6 +85,7 @@
 #define SOF_TKN_INTEL_DMIC_NUM_PDM_ACTIVE	605
 #define SOF_TKN_INTEL_DMIC_SAMPLE_RATE		608
 #define SOF_TKN_INTEL_DMIC_FIFO_WORD_LENGTH	609
+#define SOF_TKN_INTEL_DMIC_UNMUTE_RAMP_TIME     610
```
Comment (plbossart) [inline include/uapi/sound/sof/tokens.h]: unit? better to finish with e.g. RAMP_TIME_MS

File: include/sound/sof/dai-intel.h [hunk 2]
```diff
@@ -167,9 +167,10 @@ struct sof_ipc_dai_dmic_params {
 
 	uint32_t wake_up_time;      /**< Time from clock start to data (us) */
 	uint32_t min_clock_on_time; /**< Min. time that clk is kept on (us) */
+	uint32_t unmute_ramp_time;  /**< Length of logaritmic gain ramp (ms) */
```
Comment (plbossart) [inline include/sound/sof/dai-intel.h]: logarithmic?

Comment (singalsu) [reply include/sound/sof/dai-intel.h]: D'oh! My codespell isn't working apparently.

File: include/uapi/sound/sof/tokens.h (hunk 1 above)
Comment (singalsu) [reply include/uapi/sound/sof/tokens.h]: Yes, it's more clear that way.

**Comment by plbossart** [comment]
@singalsu looks good
@lgirdwood is there a matching PR on the firmware side? Are we in agreement that ABI MINOR 8 it is?

**Comment by singalsu** [comment]
@plbossart Not yet available for FW, ETA is on Monday. I spent today improving db2lin() and exp() functions accuracy those are needed on FW side to calculate the ramp.

**Comment by singalsu** [comment]
I just pushed the firmware side PR into https://github.com/thesofproject/sof/pull/1515 .

**Comment by singalsu** [comment]
It was decided to not to decibels conversion in firmware. See https://github.com/thesofproject/sof/pull/1515 .

To proceed I need your input for gain coefficient computation in in kernel in dmic load code vs. leaving it for configuration engineer's manual pen & paper calculate. In both cases this PR will change.

**Comment by plbossart** [comment]
@lgirdwood could we possibly have the conversions done by alsa-lib while processing the topology file? It looks like a generic solution that could benefit others. dB to linear and back are pretty common needs in audio.
I am not a big fan of calling an executable from M4 stuff. That's asking for trouble and adding a dependency on a new package.

**Comment by singalsu** [comment]
@plbossart @lgirdwood In the ramp time to ramp coefficient calculation the dB -> linear conversion is only the last step of the equation so having generic db2lin() function in alsa-lib is not sufficient this PR's need.

I think my two PRs set got too strict judging for code size worry, it's just 24 lines to do conversion excluding declarations vs. all the effort of doing it cleanly outside firmware. I'm pretty sure we save more in code size by e.g. limiting with SOF kconfig options the sample rates those DMIC driver currently supports (just FIR decimation factors 2 and 6 are used today for 48 kHz and 16 kHz capture - see CONFIG_CAVS_DMIC_FIR_DECIMATE_BY_x in FW code). Plus code size optimize opportunities elsewhere in SOF.

dB<->lin conversion capability has been part of firmware those I've in my earlier career worked with. Also if we would have ln() function with pretty similar Taylor series implementation we could calculate square root in firmware. Those are useful in level metering and anything attempting to process signals for humans with logarithmic perception for loudness. We can't loop such exp(), log(), sqrt() compute requests to host CPU e.g. at 10ms or 20ms from a busy algorithm processing loop because the latency would not be acceptable.

**Comment by plbossart** [comment]
@singalsu I have no objection if these functions are implemented in firmware if overall it makes sense, but the argument we should just to it because that it's simple enough doesn't fly with me. if I follow your line of thought, we could also implement the bilinear transform for IIR coefficients in firmware and just pass Q factor, center frequency and gain. That's not necessarily desirable. We should only have in firmware what is strictly required or when it simplifies the end-to-end stack.

**Comment by singalsu** [comment]
IIR design is quite tricky and have some risks, especially without double precision math, and execution time is not necessarily predictable, so I agree that such is better to avoid in firmware. FIR design can be done in contrast with a determined flow.

What I'm proposing does not have such risk. Also the implementation of the dB conversion function could be very different like table lookup & interpolation if desired. The code would still work. Especially if the header defines the fractional Q format for in & out parameters if it would vary.

We need to export some things to use space those are very specific to driver internals to end up to a correct gain coefficient value. If e.g. the internal ramp update rate would be changed to finer grid then every topology would become obsolete. I'm not saying such can't be done, but I'n confident this proposal is better to maintain.

**Pull request merged**